### PR TITLE
Add data table reporter

### PR DESCRIPTION
**Pull request by himanga**
## Description
Add a data table reporter.

## Checklist
- [x] **I am adding a new reporter.**
  - [x] I have included documentation with at least one example and description.
  - [x] I have added tests in `Tests/UnitTests/Reporters`.
  - [x] I have added a note to `CHANGELOG.md`.

## Contributing
- [x] I have read and agree to the [Contributor Agreement](https://github.com/sassoftware/jsl-hamcrest/blob/master/ContributorAgreement.txt)

Signed-off-by: Isaac Himanga

### CODE REVIEW

File: Tests/UnitTests/Reporters/DataTableReporterTest.jsl [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+ut test(DataTableReporterTests, "Success has 1 row", Expr(
+  reporter << Add Success({"", "", "label", .}, "test expr", "description", Empty());
+  ut assert value(reporter << N successes,1);
+  ut assert value(reporter << N failures,0);
+  ut assert value(reporter << N unexpected throws,0);
+  ut assert value(reporter << N total failures,0);
+));
+
+// I don't know how to write this test.
```
Comment (himanga) [inline Tests/UnitTests/Reporters/DataTableReporterTest.jsl]: From here down errors when ran recursively.  Not sure how fix..

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,8 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 
+- New data table reporter
```
Comment (emccorkle) [inline CHANGELOG.md]: Refer to it by name here with backticks: `ut data table reporter()`

File: Source/Reporters/DataTableReporter.jsl [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	New Data View = function = Method( {},
+		this:dt << New Data View;
+	);
+	
+	// Function: add expression failure
+	add expression failure = Method( {label, payload=Empty()},
+		rw = { 
+			DateTime = today(),
+			label= label,
+			test expr = "is and expression",
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl]: and -> an

File: Tests/UnitTests/Reporters/DataTableReporterTest.jsl (hunk 1 above)
Comment (emccorkle) [reply Tests/UnitTests/Reporters/DataTableReporterTest.jsl]: This works for me once a couple fixes have been done to the reporter (I'll leave comments on those).

```
ut test(DataTableReporterTests, "Returns reference", Expr(
	dtref = reporter << Get Table Reference;
	columncount = length( dtref << Get Column Names("string") );
	ut assert value(columncount , ut greater than( 4 ));
));
  
ut test(DataTableReporterTests, "Table is invisible by default", Expr(
	dtref = reporter << Get Table Reference;
	ut assert value(dtref << has data view(), 0);
));

ut test(DataTableReporterTests2, "Table is visible when requested", Expr(
	dtref = reporter << New Data View
	ut assert value(dtref << has data view(), 1);
));
```

File: Source/Reporters/DataTableReporter.jsl [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			New Column( "DateTime",        Numeric,    "Continuous", Format( "m/d/y h:m:s", 23, 0 ),
+				Input Format( "m/d/y h:m:s", 0 ), Set Values([]) ),
+			New Column( "label",           Expression, "None",    Set Values({}) ),
+			New Column( "Case Name",       Character,  "Nominal", Formula( Eval List( :label )[1] ) ),
+			New Column( "Test Name",       Character,  "Nominal", Formula( Eval List( :label )[2] ) ),
+			New Column( "Assertion Label", Character,  "Nominal", Formula( Eval List( :label )[3] ) ),
+			New Column( "Assertion ID",    Numeric,    "Nominal", Formula( Eval List( :label )[4] ) ),
+			New Column( "test expr",       Expression, "None",    Set Values({}) ),
+			New Column( "description",     Character,  "Nominal", Set Values({}) ),
+			New Column( "lre",             Expression, "None",    Set Values({}) ),
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl:40]: I think `lre` should be `Numeric` and `"Continuous"`. And all the `Empty()` below for `lre` can become `.`.

File: Source/Reporters/DataTableReporter.jsl [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+			New Column( "payload",         Expression, "None",    Set Values({}) ),
+			New Column( "result",          Character,  "Nominal", Set Values({}) )
+		);
+		1;
+	);
+
+	// Function: Table
+	// Returns a reference to the results data table
+	Get Table Reference = function = Method( {}, this:dt );
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl]: Remove the `= function` extra assignment here and below with `New Data View`. This was part of the problem with the tests.

File: Source/Reporters/DataTableReporter.jsl [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+			description = "Threw \!"" || exception || "\!"",
+			lre = Empty(),
+			payload = Name Expr( payload ),
+			result = "unexpected throw"
+		};
+		this:dt << Add Rows( rw );
+	);
+	
+	// Function: add success
+	add success = Method( {label, test expr, description, payload=Empty()},
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl:107]: It would be great if you could add an override for the new `add skip`. That was added since your first iteration of this matcher.

File: Source/Reporters/DataTableReporter.jsl [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+		};
+		this:dt << Add Rows( rw );
+	);
+
+	// Function: add failure
+	add failure = Method( { label, test expr, description, mismatch, lre, payload=Empty() },	
+		rw = { 
+			DateTime = today(),
+			label= label,
+			test expr = Expr( Expr( Name Expr( test expr ) ) ),
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl]: This can just be `Name Expr( test expr )` here and all places below.

File: Source/Reporters/DataTableReporter.jsl [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+			description = description,
+			lre = Expr( Expr( Name Expr( l ) ) ),
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl]: This can just be `lre = lre`. Other places should change to `lre = .`.

File: Source/Reporters/DataTableReporter.jsl [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+			description = "was not an expression",
+			lre = Empty(),
+			payload = Name Expr( payload ),
+			result = "expression failure"
+		};
+		this:dt << Add Rows( rw );
+	);
+
+	// Function: add failure
+	add failure = Method( { label, test expr, description, mismatch, lre, payload=Empty() },
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl:75]: I really think you should capture the `mismatch`. It is really valuable for diagnosing failures.

File: Source/Reporters/DataTableReporter.jsl [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+			DateTime = today(),
+			label= label,
+			test expr = Expr( Expr( Name Expr( test expr ) ) ),
+			description = description,
+			lre = Expr( Expr( Name Expr( l ) ) ),
+			payload = Name Expr( payload ),
+			result = "failure"
+		};
+		this:dt << Add Rows( rw );
+		wait(0);
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl]: Removing the `Wait(0)` caused me no problems and it definitely looks suspicious. Either remove it or add a comment explaining it if necessary.

**Review by emccorkle — CHANGES_REQUESTED** [review]

**Comment by emccorkle** [comment]
Also, please change the title of your pull request to remove the "WIP: " if possible.

File: Source/Reporters/DataTableReporter.jsl (hunk 4 above)
Comment (himanga) [reply Source/Reporters/DataTableReporter.jsl:40]: @emccorkle, I don't believe numeric works due to potential tests like this:
``` JSL
ut assert value( [2 2], [1 2] );
```

File: Tests/UnitTests/Reporters/DataTableReporterTest.jsl (hunk 1 above)
Comment (himanga) [reply Tests/UnitTests/Reporters/DataTableReporterTest.jsl]: Thank you @emccorkle, mistakes are so obvious once someone points them out!

**Comment by himanga** [comment]
The data/modeling type of lre is the only outstanding issue I see here. If it should be expression then I think this can be merged.

File: Source/Reporters/DataTableReporter.jsl (hunk 4 above)
Comment (emccorkle) [reply Source/Reporters/DataTableReporter.jsl:40]: You are right. Sorry about that.

File: Source/Reporters/DataTableReporter.jsl [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+			New Column( "label",           Expression, "None",    Set Values({}) ),
+			New Column( "Case Name",       Character,  "Nominal", Formula( Eval List( :label )[1] ) ),
+			New Column( "Test Name",       Character,  "Nominal", Formula( Eval List( :label )[2] ) ),
+			New Column( "Assertion Label", Character,  "Nominal", Formula( Eval List( :label )[3] ) ),
+			New Column( "Assertion ID",    Numeric,    "Nominal", Formula( Eval List( :label )[4] ) ),
+			New Column( "test expr",       Expression, "None",    Set Values({}) ),
+			New Column( "description",     Character,  "Nominal", Set Values({}) ),
+			New Column( "skipdescription", Character,  "Nominal", Set Values({}) ),
+			New Column( "lre",             Expression, "None",    Set Values({}) ),
+			New Column( "mismatch",        Expression, "None",    Set Values({}) ),
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl]: I'm pretty sure that `mismatch` should be Character.

File: Source/Reporters/DataTableReporter.jsl [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	
+	// Function: add expression failure
+	add expression failure = Method( {label, payload=Empty()},
+		rw = { 
+			DateTime = today(),
+			label= label,
+			test expr = "is an expression",
+			description = "was not an expression",
+			lre = .,
+			mismatch = .,
```
Comment (emccorkle) [inline Source/Reporters/DataTableReporter.jsl]: `mismatch` and `skip description` should be `""` rather than `.` when not relevant. Here and below.

**Review by emccorkle — CHANGES_REQUESTED** [review]
I think you are right that `lre` should be `Expression`. However, I think the data type of `mismatch` needs to change and the irrelevant values need to be changed to match the data type for a couple columns.

File: Source/Reporters/DataTableReporter.jsl (hunk 11 above)
Comment (himanga) [reply Source/Reporters/DataTableReporter.jsl]: Agreed

**Review by emccorkle — APPROVED** [review]

**Comment by emccorkle** [comment]
Thanks very much for contributing back!
